src/FsAutoComplete.Core/AdaptiveExtensions.fs

```diff
@@ -288,3 +291,517 @@ module AMap =
       >> HashMap.map (fun _ v -> AVal.constant v |> AVal.mapWithAdditionalDependenies (id))
 
     batchRecalcDirty mapping map
+
+/// <summary>
+/// A task creator that caches the task and cancels it when no longer needed.
+/// </summary>
+/// <remarks>
+/// Since the task can be references multiple times in the dependency graph, it is important to cancel it only after there are no more references to it.
+/// </remarks>
+type internal RefCountingTaskCreator<'a>(create: CancellationToken -> Task<'a>) =
```

Comment:
I added comments in the module, please ask any clarifying questions. Also @krauthaufen please let me know if I've incorrectly done something here.

Reply:
Hey, I currently have a toddler-with-feaver-problem but once this is over I'll take some time to look over that...